sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala

```diff
@@ -268,6 +268,27 @@ private[sql] abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
 
   object CartesianProduct extends Strategy {
     def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+      // Not like the equal-join, BroadcastNestedLoopJoin doesn't support condition
+      // for cartesian join, as in cartesian join, probably, the records satisfy the
+      // condition, but exists in another partition of the large table, so we may not able
+      // to eliminate the duplicates.
```

Comment:
Yes, the comment is stale. 

If we restrict the outer join condition as `None` here, then it's more like a `CartesianProduct`, that's why I put the rule in the `CartesianProduct`, and more importantly, we'd like to take those 2 rules as higher priority than the rule in Line 292.

I am totally agree with you to combine the `CartesianProduct` and `BroadcastNestedLoopJoin`, as the later just a special case of former.

Will update the code soon.